README.md
sgo - Simple Geographical Operations (with OSGB36)
================

<!-- README.md is generated from README.Rmd. Please edit that file -->
<!-- badges: start -->

[![R-CMD-check](https://github.com/clozanoruiz/sgo/workflows/R-CMD-check/badge.svg?branch=main)](https://github.com/clozanoruiz/sgo/actions?workflow=R-CMD-check)
[![coverage](https://codecov.io/gh/clozanoruiz/sgo/branch/main/graph/badge.svg?token=Qd5gkpnxFc)](https://app.codecov.io/gh/clozanoruiz/sgo?branch=main)
[![CRAN](https://www.r-pkg.org/badges/version/sgo)](https://cran.r-project.org/package=sgo)
[![GitHub release (latest
SemVer)](https://img.shields.io/github/v/release/clozanoruiz/sgo.svg?sort=semver&label=GitHub)](https://github.com/clozanoruiz/sgo/releases)
[![license](https://img.shields.io/badge/license-BSD%202--Clause-green.svg)](https://opensource.org/licenses/BSD-2-Clause)
<!-- badges: end -->

The goal of `sgo` is to provide a set of methods to transform ETRS89
coordinates to the equivalent OSGB36 National Grid coordinates, and
vice-versa, using the [Ordnance Survey’s National Grid Transformation
Model
OSTN15](https://www.ordnancesurvey.co.uk/blog/2016/09/ostn15-new-geoid-britain/).
Additionally, it also has the ability to calculate distances and areas
from groups of points defined in any of the coordinate systems
supported.

The Coordinate Systems supported by `sgo` are:

-   ETRS89: EPSGs 4258, 4937, 4936 and 3035
-   WGS84: EPSGs 4326, 4979, 4978 and 3857
-   OSGB36: EPSGs 27700, 7405 and 4277

Please note that this package assumes that the Coordinate Reference
Systems (CRS) ETRS89 and WGS84 are the same within the UK, but this
shouldn’t be a problem for most civilian use of GPS satellites. If a
high-precision transformation between WGS84 and ETRS89 is required then
it is recommended to use a different package to do the conversion.

According to the Transformations and OSGM15 User Guide, p. 8: *“…ETRS89
is a precise version of the better known WGS84 system optimised for use
in Europe; however, for most purposes it can be considered equivalent to
WGS84.”* and *“For all navigation, mapping, GIS, and engineering
applications within the tectonically stable parts of Europe (including
UK and Ireland), the term ETRS89 should be taken as synonymous with
WGS84.”*.

## Installation

You can install the released version of `sgo` from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("sgo")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("remotes")
remotes::install_github("clozanoruiz/sgo")
```

## Usage

``` r
library(sgo)

## sgo_points creation (from lists, matrices or dataframes)
sr <- sgo_points(list(-3.9369, 56.1165), epsg=4326)
sr
#> An sgo object with 1 feature (point) 
#> dimension: XY 
#> EPSG:      4326  
#>         x       y
#> 1 -3.9369 56.1165
ln <- c(-4.22472, -2.09908)
lt <- c(57.47777, 57.14965)
n <- c("Inverness", "Aberdeen")
df <- data.frame(n, ln, lt, stringsAsFactors = FALSE)
locs <- sgo_points(df, coords=c("ln", "lt"), epsg=4326)

library(maps)
map('world', regions=('uk'), xlim=c(-9, 0), ylim=c(54.5, 60.9))
points(x=sr$x, y=sr$y, pch=0, col="green")
points(locs, pch=0, col="red")
text(locs, labels=locs$n, pos=1, cex=0.9)
text(sr, labels="Stirling", pos=1, cex=0.9)
```

<img src="man/figures/README-unnamed-chunk-2-1.png" width="50%" />

``` r

## Convert coordinates to OS National Grid reference
bv <- sgo_points(list(x=247455, y=706338, name="Ben Venue"),
coords=c("x", "y"), epsg=27700)
sgo_bng_ngr(bv)$ngr
#> [1] "NN 47455 06338"
# notice the truncating:
sgo_bng_ngr(bv, digits=8)$ngr
#> [1] "NN 4745 0633"


## Convert from lon/lat to BNG
lon <- c(-4.25181,-3.18827)
lat <- c(55.86424, 55.95325)
pts <- sgo_points(list(longitude=lon, latitude=lat), epsg=4326)
sgo_lonlat_bng(pts)
#> An sgo object with 2 features (points) 
#> dimension: XY 
#> EPSG:      27700  
#>          x        y
#> 1 259174.4 665744.8
#> 2 325899.1 673996.1


## sgo_transform is a wrapper for all the conversions available.

# to BNG:
sgo_transform(locs, to=27700)
#> An sgo object with 2 features (points) and 1 field 
#> dimension: XY 
#> EPSG:      27700  
#>          x        y         n
#> 1 266698.8 845243.9 Inverness
#> 2 394104.5 806535.9  Aberdeen

# to OSGB36 (historical):
sgo_transform(locs, to=4277)
#> An sgo object with 2 features (points) and 1 field 
#> dimension: XY 
#> EPSG:      4277  
#>           x        y         n
#> 1 -4.223366 57.47804 Inverness
#> 2 -2.097450 57.14985  Aberdeen

# to Pseudo-Mercator:
sgo_transform(locs, to=3857)
#> An sgo object with 2 features (points) and 1 field 
#> dimension: XY 
#> EPSG:      3857  
#>           x       y         n
#> 1 -470293.7 7858404 Inverness
#> 2 -233668.5 7790768  Aberdeen


## Convert OS National Grid references to ETRS89
munros <- data.frame(
  ngr=c("NN 1652471183", "NN 98713 98880",
         "NN9521499879", "NN96214 97180"),
  name=c("Beinn Nibheis", "Beinn Macduibh",
          "Am Bràigh Riabhach", "Càrn an t-Sabhail"))
sgo_transform(sgo_ngr_bng(munros, col="ngr"), to=4258)
#> An sgo object with 4 features (points) and 1 field 
#> dimension: XY 
#> EPSG:      4258  
#>           x        y               name
#> 1 -5.005929 56.79589      Beinn Nibheis
#> 2 -3.672139 57.06976     Beinn Macduibh
#> 3 -3.730236 57.07795 Am Bràigh Riabhach
#> 4 -3.712631 57.05394  Càrn an t-Sabhail


## Calculate distances
# Distance in metres from one point to 2 other points
p1 <- sgo_points(list(-3.9369, 56.1165), epsg=4326)
lon <- c(-4.25181,-3.18827)
lat <- c(55.86424, 55.95325)
pts <- sgo_points(list(longitude=lon, latitude=lat), epsg=4326)
p1.to.pts <- sgo_distance(p1, pts, by.element = TRUE)
p1.to.pts
#> [1] 34279.98 50081.63

# Perimeter in metres of a polygon defined as a series of ordered points:
lon <- c(-6.43698696, -6.43166843, -6.42706831, -6.42102546,
-6.42248238, -6.42639092, -6.42998435, -6.43321409)
lat <- c(58.21740316, 58.21930597, 58.22014035, 58.22034112,
58.21849188, 58.21853606, 58.21824033, 58.21748949)
pol <- sgo_points(list(lon, lat), epsg=4326)
# Let's create a copy of the polygon with its coordinates shifted one position
# so that we can calculate easily the distance between vertices
coords <- sgo_coordinates(pol)
pol.shift.one <- sgo_points(rbind(coords[-1, ], coords[1, ]), epsg=pol$epsg)
perimeter <- sum(sgo_distance(pol, pol.shift.one, by.element=TRUE))
perimeter
#> [1] 2115.33


## Area of an ordered set of points
A <- sgo_area(pol)
sprintf("%1.2f", A)
#> [1] "133610.63"
# Interpolate new vertices (here every 10 metres) if more accuracy is needed
A <- sgo_area(pol, interpolate=10)
sprintf("%1.2f", A)
#> [1] "133610.64"
```

## Benchmark

Very simple benchmark in a Intel(R) Core(TM) i7-6700HQ CPU @2.60GHz, and
16 GB of RAM. See the code of this README file to see the R statements
used.

<img src="man/figures/README-unnamed-chunk-4-1.png" width="100%" />


NEWS.md
# sgo 0.9.2

- Bumped minimum R required version to 3.5 (sysdata.rda uses workspace format version 3)
- Replaced `testthat` with `tinytest` for Unit Testing.
- sgo_coordinates gained a couple of paramaters meant for geodetic coordinates.
- Improved the speed of NGR conversions.
- A little tidy-up of both code and documentation.


# sgo 0.9.1

- **Fixed**: Now, when converting from 2D BNG to 3D BNG, the input is first converted  
  to 3D  by filling z with zeros. Previous versions just added zeros to the output.
- **Fixed**: When converting from 2D to 3D coordinates with the same datum the z  
  column was not  being filled with zeroes.
- **Fixed**: Points right on the upper boundaries of the OSTN grid are considered   
  outside of the OSTN15 rectangle.
- **Fixed**: Output text of the print function.
- Improved slightly the speed of 3D calculations from/to BNG.
- Improved some parts of sgo_transforms's documentation.


# sgo 0.9.0

- Initial release on CRAN.


R/points.R
#################################################################################
#                                                                               #
# Version:  1.0                                                                 #
# Revision: 0 - 26/02/2018. Published version                                   #
#                                                                               #
#################################################################################

#' @encoding UTF-8
#' @title Object containing 2D point coordinates
#'
#' @description
#' 2D coordinates (and other atributes) of a point or collection of points
#'
#' @name sgs_points
#' @usage sgs_points(x, coords=NULL, epsg = NULL)
#' @param x A list or dataframe with at least 2 columns of either easting/northing
#' or latitude/longitude coordinates per row - depending on \code{epsg}. \strong{Please
#' note} that the order is important: lon/lat or northing/easting will produce
#' wrong results.
#'
#' It also accepts objects of class \code{sf POINT/MULTIPOINT} or
#' \code{sfc POINT/MULTIPOINT}.
#' @param coords A vector with the names of the columns containing the easting/northing
#' (or latitude/longitude) coordinates.
#' @param epsg Specifies the EPSG code of coordinates to store. It can take any
#' of the following values: \code{4326}, \code{3857}, \code{4277}, \code{27700} or \code{4258}.
#' @details
#' This object stores 2D point coordinates and any other column-list of attributes
#' related to each point. Currently it only supports 5 \code{epsg}s
#' for coordinates:
#' \itemize{
#' \item\code{4326}: WGS84, geodetic coordinate system. The 2 columns in \code{x} be defined
#' as Latitude and Longitude values respectively. The defined datum for this set of
#' coordinates is WGS84 (https://epsg.io/4326)
#'
#' \item\code{3857}: WGS 84 / Pseudo-Mercator, projected coordinate system. The 2 columns
#' in \code{x} must be defined as Easting and Northing values respectively. The defined
#' datum for this set of coordinates is WGS84 (https://epsg.io/3857)
#'
#' \item\code{4277}: OSGB36, geodetic coordinate system. The 2 columns in \code{x}
#' must be defined as Latitude and Longitude values respectively. The defined datum for this
#' set of coordinates is OSGB 1936 (https://epsg.io/4277). \emph{Coordinates defined
#' like this should only be used for historical reasons.}
#'
#' \item\code{27700}: British National Grid, projected coordinate system. The 2 columns
#' in \code{x} must be defined as Easting and Northing values respectively. The defined
#' datum for this set of coordinates is OSGB 1936 (https://epsg.io/27700).
#'
#' \item\code{4258}: ETRS89, geodetic coordinate system. The 2 columns in \code{x}
#' must be defined as Latitude and Longitude values respectively. The defined
#' datum for this set of coordinates is ETRS89 (https://epsg.io/4258).
#' }
#' @return
#' An object of class \code{sgs_points}.
#' @seealso \code{\link{sgs_polygons}}, \code{\link{sgs_points_xy}},
#' \code{\link{sgs_points_sf}}, \code{\link{sgs_set_gcs}}.
#' @examples
#' # lists:
#' p1 <- sgs_points(list(56.1165, -3.9369), epsg=4326)
#' lat <- c(55.86424, 55.95325)
#' lon <- c(-4.25181,-3.18827)
#' p2 <- sgs_points(list(latitude=lat, longitude=lon), epsg=4326)
#' # dataframe:
#' lt <- c(57.47777, 57.14965)
#' ln <- c(-4.22472, -2.09908)
#' n <- c("Inverness", "Aberdeen")
#' df <- data.frame(n, lt, ln, stringsAsFactors = FALSE)
#' p3 <- sgs_points(df, coords=c("lt", "ln"), epsg=4326)
#'
#' \dontrun{
#' library(maps)
#' map('world', regions=('uk'))
#' points(x=p1$longitude, y=p1$latitude, pch=0, col="green")
#' points(x=p2$lon, y=p2$lat, pch=0, col="blue")
#' points(x=p3$lon, y=p3$lat, pch=0, col="red")
#' }
#' @export
sgs_points <- function (x, coords=NULL, epsg=NULL) UseMethod("sgs_points")

#' @export
sgs_points.list <- function (x, coords=NULL, epsg=NULL) {

  # Checks
  len <- length(x)
  if (len < 2) stop("This method accepts lists with at least 2 columns")

  if(is.null(epsg) || (!epsg %in% epsgs[, "epsg"])) {
    stop("The epsg parameter must be entered as 4326, 3857, 4277, 27700 or 4258")
  }

  if (len == 2 && is.null(names(x)) ) {
    coords <- coordinates.names[1, ] #just because
    names(x) <- coords
  }

  if(is.null(coords)) {
    lnames <- tolower(names(x))
    known.coords <- t(apply(coordinates.names, 1, function(n) n %in% lnames))
    if (any(known.coords)) {
      coords <- names(x)[lnames %in% coordinates.names[known.coords]]
    } else { stop("Must specify the name of columns containing coordinates using the 'coords' paramater") }
  }

  if(!is.numeric(x[[coords[1]]]) | !is.numeric(x[[coords[2]]]))
    stop("All coordinates must be numeric")

  other.columns <- x[!(names(x) %in% coords)]
  if (length(other.columns)==0) other.columns <- NULL
  points <- switch(as.character(epsg),
                   "4326" = structure(c(other.columns,
                                        list(latitude = x[[coords[1]]], longitude = x[[coords[2]]],
                                             epsg=epsg, datum="WGS84")),
                                      class="sgs_points"),
                   "3857" = structure(c(other.columns,
                                         list(easting = x[[coords[1]]], northing = x[[coords[2]]],
                                              epsg=epsg, datum="WGS84")),
                                       class="sgs_points"),
                   "4277" = structure(c(other.columns,
                                        list(latitude = x[[coords[1]]], longitude = x[[coords[2]]],
                                             epsg=epsg, datum="OSGB36")),
                                      class="sgs_points"),
                   "27700" = structure(c(other.columns,
                                        list(easting = x[[coords[1]]], northing = x[[coords[2]]],
                                             epsg=epsg, datum="OSGB36")),
                                      class="sgs_points"),
                   "4258" = structure(c(other.columns,
                                         list(latitude = x[[coords[1]]], longitude = x[[coords[2]]],
                                              epsg=epsg, datum="ETRS89")),
                                       class="sgs_points")
  )

}

#' @export
sgs_points.data.frame <- function (x, coords=NULL, epsg=NULL) {

  # Checks
  cols <- ncol(x)
  if (cols < 2) stop("This method accepts dataframes with at least 2 columns")

  if(is.null(epsg) || (!epsg %in% epsgs[, "epsg"])) {
    stop("The epsg parameter must be entered as 4326, 3857, 4277, 27700 or 4258")
  }

  if (cols == 2) {
    coords <- coordinates.names[1, ] #just because
    names(x) <- coords #when there are only 2 cols, which names we apply shouldn't matter
  }

  sgs_points(as.list(x), coords=coords, epsg=epsg)

}

#' @export
sgs_points.sf <- function (x, coords=NULL, epsg=NULL) {

  epsg <- sf_checks(x, epsg)

  coordinates <- sf::st_coordinates(x)
  coords <- colnames(coordinates)
  rownames(coordinates) <- NULL

  sf::st_geometry(x) <- NULL
  sgs_points(as.list(cbind(x, coordinates)), coords=coords, epsg=epsg)

}

#' @export
sgs_points.sfc <- function (x, coords=NULL, epsg=NULL) {

  epsg <- sf_checks(x, epsg)

  coordinates <- sf::st_coordinates(x)
  coords <- colnames(coordinates)
  rownames(coordinates) <- NULL

  lst <- list(coordinates[, 1], coordinates[, 2])
  names(lst) <- coords
  sgs_points(lst, coords=coords, epsg=epsg)

}

sf_checks <-function (x, epsg) {

  # Check if input data is a simple feature object (sf, sfc).
  # We need sf to get the coordinates of the object
  if (!requireNamespace("sf", quietly = TRUE)) {
    stop("Package \"sf\" needed for this function to work. Please install it.",
         call. = FALSE)
  }

  if (all(!sf::st_is(x, c("POINT", "MULTIPOINT")))) stop("Only POINT and MULTIPOINT geometries are accepted")

  if (is.null(epsg)) {
    crs.epsg <- sf::st_crs(x)$epsg
    if(!crs.epsg %in% epsgs[, "epsg"]) {
      stop("The parameter epsg must be entered, or an appropiate crs defined for the sf object")
    } else {
      epsg <- crs.epsg
    }
  }

  epsg

}

#' @encoding UTF-8
#' @title Convert sgs_points objects to sf objects
#'
#' @description
#' Converts \code{sgs_points} objects to instances of class \code{sf}. \pkg{sf}
#' is a package that allows working with spatial vector data.
#'
#' The reverse conversion (from \code{sf} to \code{sgs_points}) is done
#' automatically by the \code{\link{sgs_points}} contructor.
#'
#' @name sgs_points_sf
#' @usage sgs_points_sf(x)
#' @param x An instance of \code{sgs_points}.
#' @return
#' An object of class \code{sf}.
#' @examples
#' p <- sgs_points(list(57.47777, -4.22472), epsg=4326)
#' sf.p <- sgs_points_sf(p)
#' sf::st_crs(sf.p)
#' @export
sgs_points_sf <- function (x) UseMethod("sgs_points_sf")

#' @export
sgs_points_sf.sgs_points <-function(x) {

  x <- unclass(x)
  names <- names(x)
  columns <- !names %in% c("epsg", "datum")

  # Projection type
  coord.system <- epsgs[epsgs[, "epsg"]==x$epsg, "type"]

  if(coord.system == "GCS") {
    coords <- c("longitude", "latitude")
    len.coords <- length(x$latitude)
  } else {
    coords <- c("easting", "northing")
    len.coords <- length(x$easting)
  }

  # Ensure all extra columns have the same numbers of elements as coordinates
  # This means there won't be 'recycling' of elements
  x[columns] <- lapply(x[columns], function(col, len) {
                                    length(col) <- len; col}, len=len.coords)

  sf::st_as_sf(data.frame(x[columns], stringsAsFactors = FALSE),
               coords=coords, crs =x$epsg, dim="XY")

}

#' @encoding UTF-8
#' @title Extracts coordinates from an \code{sgs_points} object
#'
#' @description
#' Extract the coordinates of an \code{sgs_points} object expressed as a matrix.
#'
#' @name sgs_points_xy
#' @usage sgs_points_xy(x)
#' @param x An instance of \code{sgs_points}.
#' @return
#' A matrix with 2 named columns.
#' @examples
#' p <- sgs_points(list(57.47777, -4.22472), epsg=4326)
#' coords <- sgs_points_xy(p)
#' @export
sgs_points_xy <- function (x) UseMethod("sgs_points_xy")

#' @export
sgs_points_xy.sgs_points <-function(x) {

  # Projection type
  coord.system <- epsgs[epsgs[, "epsg"]==x$epsg, "type"]

  coords <- if(coord.system == "GCS") {
    c("latitude", "longitude")
  } else {
    c("easting", "northing")
  }

  xy <- matrix(unlist(x[coords]), ncol=2)
  colnames(xy) <- coords

  xy

}

# Extending '[' function to support sgs_points:
# Attributes are usually lost when subsetting lists, therefore we need to
# extend a subsetting operator that will keep all attributes of our object.
`[.sgs_points` <- function(x, i, ...) {

  if (epsgs[epsgs[, "epsg"]==x$epsg, "type"] == "GCS") {
    core.cols <- sgs_points.core[!sgs_points.core %in% c("easting", "northing")]
  } else {
    core.cols <- sgs_points.core[!sgs_points.core %in% c("latitude", "longitude")]
  }

  r <- NextMethod("[")
  if ( all(core.cols %in% names(r)) ) class(r) <- "sgs_points"
  #if (!inherits(r, "sgs_points")) class(r) <- "sgs_points"
  r

}

# TODO print!
print.sgs_points <- function(o) {
  cat(o$epsg, "\n")
  cat(o$datum, "\n")
}

# sgs_equal.sgs_points
# BETTER DO THIS WITH SF (AND MOSTLY EVERY OTHER GEO FUNCTION)!!!
